### PR TITLE
WIP [helpers] email_case_deletion() format error fixed

### PR DESCRIPTION
**Pull request by gitaarik**
**NOTE: This merge request is work in progress**

`email_case_deletion()` didn't work because of wrong string formatting, I fixed this

I also added a unit tests for this, but it doesn't work because I get an error in rendering the notification email template:

```
ERROR: test_send_mail_to_case_author (tcms.testcases.tests.test_models.TestSendMailOnCaseIsDeleted)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/mock/mock.py", line 1305, in patched                                                                                           
    return func(*args, **keywargs)
  File "/home/rik/dev/kiwi/tcms/testcases/tests/test_models.py", line 188, in test_send_mail_to_case_author                                                                                                         
    expected_body = render_to_string('email/post_case_delete/email.txt', {'case': self.case})
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/loader.py", line 61, in render_to_string                                                                       
    template = get_template(template_name, using=using)
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/loader.py", line 15, in get_template                                                                           
    return engine.get_template(template_name)
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/backends/django.py", line 34, in get_template                                                                  
    return Template(self.engine.get_template(template_name), self)
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/engine.py", line 144, in get_template                                                                          
    template, origin = self.find_template(template_name)
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/engine.py", line 126, in find_template                                                                         
    template = loader.get_template(name, skip=skip)
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/loaders/base.py", line 30, in get_template                                                                     
    contents, origin, origin.template_name, self.engine,
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/base.py", line 156, in __init__                                                                                
    self.nodelist = self.compile_nodelist()
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/base.py", line 194, in compile_nodelist                                                                        
    return parser.parse()
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/base.py", line 478, in parse                                                                                   
    raise self.error(token, e)
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/base.py", line 476, in parse                                                                                   
    compiled_result = compile_func(self, token)
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/templatetags/i18n.py", line 465, in do_block_translate                                                                  
    value = token_kwargs(remaining_bits, parser, support_legacy=True)
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/base.py", line 1039, in token_kwargs                                                                           
    kwargs[key] = parser.compile_filter(value)
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/base.py", line 563, in compile_filter                                                                          
    return FilterExpression(token, self)
  File "/home/rik/.local/share/virtualenvs/kiwi-_tsNZa2y/lib/python3.6/site-packages/django/template/base.py", line 663, in __init__                                                                                
    "from '%s'" % (token[upto:], token))
django.template.exceptions.TemplateSyntaxError: Could not parse the remainder: '().history_user.username' from 'case.history.latest().history_user.username'
```

It seems that `case.history.latest().history_user` is `None`. I looked at the `TestSendMailOnCaseIsUpdated` test and there the email is sent differently, it's not using a template and it has a fallback to an empty string:

https://github.com/kiwitcms/Kiwi/blob/master/tcms/core/history.py#L54

So I'm not sure what's best to do here. I think a template for such a simple message as:

https://github.com/kiwitcms/Kiwi/blob/master/tcms/templates/email/post_case_delete/email.txt

Is maybe a bit overdone, and it would make it easier to do it the same way like the update case email notification is done.

What do you guys think?

### CODE REVIEW

File: tcms/locale/de_DE/LC_MESSAGES/django.po [hunk 1]
```diff
@@ -49,6 +49,21 @@ msgstr "URL"
 msgid "Comment"
 msgstr "Bemerkung"
 
+#: tcms/core/contrib/linkreference/tests.py:84
+msgid "Enter a valid URL."
+msgstr ""
```
Comment (atodorov) [inline tcms/locale/de_DE/LC_MESSAGES/django.po]: Use `make messages` to regenerate .po files. They should exclude translations in tests.

**Comment by atodorov** [comment]
For the failing tests see `kiwi_auth/tests.py` or `testcases/tests/test_models.py` right at the end of the file. I think you should try to delete the test case and assert that send_mail was called down the line.

**Comment by gitaarik** [comment]
Ran `./manage.py makemessages` and updated the pull request

**Comment by gitaarik** [comment]
The test case doesn't fail, it errors. Because of the template error I 
described. Did you read my full description?

**Comment by atodorov** [comment]
> Did you read my full description?

Yes, did you read mine ? It shows you where to find a code snippet which you can reuse for your test.

**Comment by gitaarik** [comment]
Yes, I am trying to call assert that send_mail was called down the line, see:

https://github.com/kiwitcms/Kiwi/pull/878/files#diff-dcd007cd20d7bd1dbed6ddc7d0237fb3R194

I am trying to make a similar unit test to the one in `testcases/tests/test_models.py` `TestSendMailOnCaseIsUpdated` as you pointed out. But unfortunately it's not possible to do it as similar in the current situation, because the way the template for the notification email for delete operations is build is different from the way it is done at the update operation. That's why I'm asking you guys what the best thing is to do here. Shall we change the way it is done in the delete operation to match the update operation or the other way around?

**Comment by atodorov** [comment]
hmmm, history is missing and I have the feeling some signals don't get fired. Nothing immediately obvious. There is `@factory.django.mute_signals(signals.post_save)` for `TestCaseFactory` and indeed history is saved in post_save but removing this didn't solve the issue.

This will take some debugging, can't figure it out at the top of my head.

**Comment by gitaarik** [comment]
I'm not following you, what does this have to do with muting signals?

**Comment by atodorov** [comment]
@gitaarik I am still behind on some tasks and will take me some time to figure out what is going on here. There may be a bug somewhere wrt how history gets passed to the object or just not available during testing. Can't say ATM.

Your patch looks good though so do you mind splitting it into a separate commit. I can cherry-pick it by hand or merge another PR. And then I can work on the test.

**Comment by gitaarik** [comment]
Hey @atodorov, take your time. I am willing to put effort into fixing the issue, I just need an opinion from somebody who is more familiar with the code. When you have the time please read through the issue and tell me what you think.

**Comment by gitaarik** [comment]
Because as I understand it, there's just a difference in the way the template for the delete notification email is build compared to the template for the update notification email. And currently there is no unit test that checks the delete notification email, but there is one for the update notification email. Now I am trying to add that unit test, but I am unable to do it the same way as the unit test for the update notification email because the template is build differently.

See the difference here:

[email_case_update()](https://github.com/kiwitcms/Kiwi/blob/master/tcms/testcases/helpers/email.py#L8-L14)

[email_case_deletion()](https://github.com/kiwitcms/Kiwi/blob/master/tcms/testcases/helpers/email.py#L17-L25)

**Comment by atodorov** [comment]
I have cherry-picked only the fix   without .po files changes and without the test. You can rebase whn you have time. I'll try to see why this test fails over the weekend.

**Comment by atodorov** [comment]
@gitaarik the latest master fixes the TemplateSyntaxError you reported in comment #0. 

It's been a while but please rebase this PR and leave only the tests. The rest has been included already.

**Pull request closed without merging**